src/Symfony/Component/Validator/Test/CompoundConstraintTestCase.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Test;
+
+use PHPUnit\Framework\ExpectationFailedException;
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\Constraints\Compound;
+use Symfony\Component\Validator\Constraints\CompoundValidator;
+use Symfony\Component\Validator\ConstraintViolationListInterface;
+use Symfony\Component\Validator\Context\ExecutionContext;
+use Symfony\Component\Validator\Context\ExecutionContextInterface;
+use Symfony\Component\Validator\Validation;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+/**
+ * A test case to ease testing Compound Constraints.
+ *
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ */
+abstract class CompoundConstraintTestCase extends TestCase
+{
+    protected ValidatorInterface $validator;
+    protected ?ConstraintViolationListInterface $violationList = null;
+    protected ExecutionContextInterface $context;
+    protected string $root;
+
+    private mixed $validatedValue;
+
+    protected function setUp(): void
+    {
+        $this->root = 'root';
+        $this->validator = $this->createValidator();
+        $this->context = $this->createContext($this->validator);
+    }
+
+    protected function validateValue(mixed $value): void
+    {
+        $this->validator->inContext($this->context)->validate($this->validatedValue = $value, $this->createCompound());
+    }
+
+    protected function createValidator(): ValidatorInterface
+    {
+        return Validation::createValidator();
+    }
+
+    protected function createContext(?ValidatorInterface $validator = null): ExecutionContextInterface
+    {
+        $translator = $this->createMock(TranslatorInterface::class);
+        $translator->expects($this->any())->method('trans')->willReturnArgument(0);
+
+        return new ExecutionContext($validator ?? $this->createValidator(), $this->root, $translator);
+    }
+
+    public function assertViolationsRaisedByCompound(Constraint|array $constraints): void
+    {
+        if ($constraints instanceof Constraint) {
+            $constraints = [$constraints];
+        }
+
+        $validator = new CompoundValidator();
+        $context = $this->createContext();
+        $validator->initialize($context);
+
+        $validator->validate($this->validatedValue, new class($constraints) extends Compound {
+            public function __construct(public array $constraints)
```

Comment:
You redeclare and assign the `$constraints` property to use it in the `getConstraints()` implementation which is called by the parent class to populate the `$constraints` property. This is very confusing.

I suggest we find a different name for a local `$constraints` property and make that property private.

Reply:
Got it. What do you think of this new naming, `$testedConstraints`?

Reply:
👌🏻